CppLinq/cpplinq.hpp

```diff
@@ -3788,6 +4161,102 @@ namespace cpplinq
         };
 
         // -------------------------------------------------------------------------
+
+        template<typename TPredicate>
+        struct last_predicate_builder : base_builder
+        {
+            typedef                 last_predicate_builder<TPredicate>     this_type           ;
+            typedef                 TPredicate                              predicate_type      ;
+
+            predicate_type          predicate   ;
+
+            CPPLINQ_INLINEMETHOD last_predicate_builder (predicate_type predicate) CPPLINQ_NOEXCEPT
+                :   predicate (std::move (predicate))
+            {
+            }
+
+            CPPLINQ_INLINEMETHOD last_predicate_builder (last_predicate_builder const & v) CPPLINQ_NOEXCEPT
+                :   predicate (v.predicate)
+            {
+            }
+
+            CPPLINQ_INLINEMETHOD last_predicate_builder (last_predicate_builder && v) CPPLINQ_NOEXCEPT
+                : predicate (std::move (v.predicate))
+            {
+            }
+
+            template<typename TRange>
+            CPPLINQ_INLINEMETHOD typename TRange::value_type build (TRange range)
+            {
+                bool found = false;
+
+                auto current = typename TRange::value_type ();
+
+                while (range.next ())
+                {
+                    if (predicate (range.front ()))
+                    {
+                        found = true;
+                        current = std::move (range.front ());
+                    }
+                }
+
+                if (found)
+                {
+                    return current;
+                }
+                else
+                {
+                    throw sequence_empty_exception ();
+                }
+            }
+
+        };
+
+        // -------------------------------------------------------------------------
+
+        struct last_builder : base_builder
+        {
+            typedef                 last_builder                   this_type       ;
+
+            CPPLINQ_INLINEMETHOD last_builder () CPPLINQ_NOEXCEPT
+            {
+            }
+
+            CPPLINQ_INLINEMETHOD last_builder (last_builder const & v) CPPLINQ_NOEXCEPT
+            {
+            }
+
+            CPPLINQ_INLINEMETHOD last_builder (last_builder && v) CPPLINQ_NOEXCEPT
+            {
+            }
+
+            template<typename TRange>
+            CPPLINQ_INLINEMETHOD typename TRange::value_type build (TRange range)
+            {
+                bool found = false;
+
+                auto current = typename TRange::value_type ();
```

Comment:
I don't know how .first looks but this requires value_type to be default-constructible. What about using opt to avoid this?